### PR TITLE
(#341, #462, #187)  Refactor Pager Component / PN bug fix

### LINKED ISSUE #341: Refactor CTS Pager component
### As a developer, refactor the CTS pager component so that it does not cause unnecessary rerenders and is more performant.

> **ESTIMATE** 5

---

## Solution
Implement the most recent form of the pager from react-playground, or one of the more recent projects using the pager.

### Technical Notes
- The current CTS pager is based on the R4R pager and is very closely intertwined with the redux global store and results "caching" scheme
- Remove dependence on redux stores and middleware
- Eliminate unnecessary re-renders (using the pager should only cause rerenders of the results list and supporting bits, not the entire results page
- Fix any console warnings

### LINKED ISSUE #462: Blank page displays intermittently when clicking/copying a URL
### Issue description
So this is a weird, intermittent issue.

A client reported that when she clicked (or copied, not sure) [this URL](https://www.cancer.gov/about-cancer/treatment/clinical-trials/search/r?d=C923%7CC15347&loc=0&pn=2&rl=2&t=C2955&tt=treatment) from a previous search, she got a blank page. Another client reported that the trial results loaded correctly for her. When I clicked or copied the URL, I was most often redirected to [this URL](https://www.cancer.gov/about-cancer/treatment/clinical-trials/search/r?d=C923%7CC15347&loc=0&pn=3&rl=2&t=C2955&tt=treatment) (with a different pn query param than the original URL), but the trial results loaded for that page number. Very intermittently, I was redirected [here](https://www.cancer.gov/about-cancer/treatment/clinical-trials/search/r?d=C923%7CC15347&loc=0&pn=3&rl=2&t=C2955&tt=treatment) again, but then saw a blank page.

> **ESTIMATE** TBD

### Steps to reproduce the issue

1.  Go to https://www.cancer.gov/about-cancer/treatment/clinical-trials/search/r?d=C923%7CC15347&loc=0&pn=2&rl=2&t=C2955&tt=treatment
2. Either:
   1. The page redirects to the third page of results and shows the trial results
   2. The page redirects to the third page of results and shows a blank page

### What's the expected result?
- The page should not redirect, and show the second page of results with the trial results

### What's the actual result?
-  Either:
   1. The page redirects to https://www.cancer.gov/about-cancer/treatment/clinical-trials/search/r?d=C923%7CC15347&loc=0&pn=3&rl=2&t=C2955&tt=treatment and shows the trial results for that page number
   2. The page redirects to https://www.cancer.gov/about-cancer/treatment/clinical-trials/search/r?d=C923%7CC15347&loc=0&pn=3&rl=2&t=C2955&tt=treatment and shows a blank page

### Additional details / screenshot
image002

### PR DESCRIPTION
**Pull request by adriancofie**
(#341, #462, #187)  Refactor Pager Component / PN bug fix

 ` /src/components/atomic/Pager`
  `/src/views/ResultsPage/`

* Replaces the existing Pager component with the Pager from CTL

- Simplifies the logic for the pager component
   - Uses absolute value to calculated distance instead of max/min combination
   - Renames variables for clarity
   - Restructures ellipsis logic for clarity
   -  See: https://github.com/NCIOCPL/clinical-trials-listing-app/pull/146/ for direct comparison

* Updates ResultsPage.jsx to use the new pager
* Updates ResultsListItem to remove the addition of 1 to the analytics page number
* Updates Pager currentPageNeighbors to 2
      
  --------

  `src/utilities/__tests__/`
 ` cypress/integration/`

* Updates tests to account for removal of Edge case in new Pager.
* Updates integration tests to account for new pager class names
* Updates unit tests to account for defaultiing to 1 as the starting page instead of 0
* Adds cyrpress scenario for forward and back browser navigation
* Updates tests to account for currentPageNeighbors req change
   
   ---------

 ` src/utilities`
 ` src/store/reducers/form.js`
 ` src/constants/index.js`

   *   Resolves bugs related to directly navigating to a search url/pn
   *   Resolves bugs related to auto-incrementing page number
   *   Adds lines for complaining linter rule inherited from testing-library
   ---------

* Closes #341
* Closes #462

### CODE REVIEW

File: src/utilities/__tests__/formatTrialSearchQuery.js [hunk 1]
```diff
@@ -537,7 +538,7 @@ function getPhaseObject(phases) {
 }
 
 /**
- * Helper function to set enabled trial types.
+ * Helper function to set enab led trial types.
```
Comment (seyilonge-nci) [inline src/utilities/__tests__/formatTrialSearchQuery.js:541]: The extra space on "enabled" must have slipped through.

**Review by seyilonge-nci — APPROVED** [review]

**Comment by kate-mashkina** [comment]
2 questions for @sarinapadilla 
1) The ticket says we are making CTS pager off listing app pager example - does this also mean we are changing the neighboring pages logic to be same with CTL? Below is screenshot for the reference, but in short CTS shows 1 neighbor and CTL shows 2 pages
Screen Shot 2022-02-28 at 2 04 20 PM

2) Should we address the case when non-existent page number provided ? Currently app just keeps loading and I am not sure if this is a desired behavior 
Screen Shot 2022-02-28 at 2 09 51 PM

**Comment by kate-mashkina** [comment]
Hitting browser back or forward arrows changes the url (pagination) but not re-loading the needed page

**Comment by sarinapadilla** [comment]
@kate-mashkina @adriancofie addressing your questions:
1. Yes, we are changing the neighboring pages logic to be the same as CTL. CTS should now show 2 neighboring pages rather than 1.
2. The app should error when given a non-existent page number. Is it possible to display the invalid criteria error message with the page number listed as the invalid field?

**Comment by adriancofie** [comment]
Added the following scenario: 
```

	Scenario: user can navigate using the browser forward and back buttons
		Given the user navigates to "/about-cancer/treatment/clinical-trials/search/r?loc=0&rl=1"
		Then the page title is "Clinical Trials Search Results"
		And result list is displayed
		And pager displays the following navigation options
			| pages  |
			| 1      |
			| 2      |
			| 3      |
			| ...    |
			| 683    |
			| Next > |
		And the page "1" is highlighted
		When user clicks on "Next >" button
		Then the user is redirected to "/about-cancer/treatment/clinical-trials/search/r" with query parameters "loc=0&pn=2&rl=1"
		And result list is displayed
		When user clicks the browser back button
		Then the user is redirected to "/about-cancer/treatment/clinical-trials/search/r" with query parameters "loc=0&rl=1"
		And result list is displayed
		And pager displays the following navigation options
			| pages  |
			| 1      |
			| 2      |
			| 3      |
			| ...    |
			| 683    |
			| Next > |
		When user clicks the browser forward button
		Then the user is redirected to "/about-cancer/treatment/clinical-trials/search/r" with query parameters "loc=0&pn=2&rl=1"
		And result list is displayed
		And pager displays the following navigation options
			| pages      |
			| < Previous |
			| 1          |
			| 2          |
			| 3          |
			| 4          |
			| ...        |
			| 683        |
			| Next >     |
```

**Review by kate-mashkina — APPROVED** [review]

**Comment by sarinapadilla** [comment]
@adriancofie does this PR also address #187?

**Comment by adriancofie** [comment]
Yep!
